flocker/control/_persistence.py

```diff
@@ -164,6 +164,32 @@ def upgrade_from_v3(cls, config):
         }
         return dumps(decoded_config)
 
+    @classmethod
+    def upgrade_from_v4(cls, config):
+        """
+        Migrate a v4 JSON configuration to v5.
+
+        :param bytes config: The v4 JSON data.
+        :return bytes: The v5 JSON data.
+        """
+        decoded_config = loads(config)
+        decoded_config[u"version"] = 5
+        try:
+            nodes = decoded_config[u"deployment"][u"nodes"]
+        except KeyError:
+            pass
+        else:
+            new_nodes = []
+            for n in nodes:
+                new_node = n
+                new_node[u"applications"] = {
+                    u"values": [(a[u"name"], a) for a in n[u"applications"]],
+                    _CLASS_MARKER: "PMap"
```

Comment:
Agreed that it should be, unfortunately that's not the way our wire_encoding works :(

https://github.com/ClusterHQ/flocker/blob/d0f65ca210a23cc3152c5ca68c1c35ae1c1d8659/flocker/control/_persistence.py#L191

(Also see other upgraders in this file)

Bonus points: Because this is in the mechanism that we use to load configs generated from previous versions it would be quite a bit of work to change this now :(.